<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Modules · NextGP.jl</title><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="home/"><img src="assets/logo.png" alt="NextGP.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="home/">NextGP.jl</a></span></div><form class="docs-search" action="search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="home/">Home</a></li><li class="is-active"><a class="tocitem" href>Modules</a><ul class="internal"><li><a class="tocitem" href="#Basic-Model"><span>Basic Model</span></a></li><li><a class="tocitem" href="#Public-Functions"><span>Public Functions</span></a></li><li><a class="tocitem" href="#Internals"><span>Internals</span></a></li><li><a class="tocitem" href="#Convenience-functions"><span>Convenience functions</span></a></li></ul></li><li><span class="tocitem">Examples</span><ul><li><a class="tocitem" href="Example/Example/">Example</a></li><li><a class="tocitem" href="PBLUP/PBLUP/">PBLUP</a></li><li><a class="tocitem" href="BWGR/BWGR/">BWGR</a></li><li><a class="tocitem" href="MultipleMarkerSets/MultipleMarkerSets/">Multiple marker sets</a></li><li><a class="tocitem" href="BayesLV/BayesLV/">BayesLV</a></li></ul></li><li><a class="tocitem" href="citation/">Citation</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Modules</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Modules</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/datasciencetoolkit/NextGP.jl/blob/main/docs/src/index.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="NextGP.jl:-Modules-and-Functions"><a class="docs-heading-anchor" href="#NextGP.jl:-Modules-and-Functions">NextGP.jl: Modules and Functions</a><a id="NextGP.jl:-Modules-and-Functions-1"></a><a class="docs-heading-anchor-permalink" href="#NextGP.jl:-Modules-and-Functions" title="Permalink"></a></h1><ul><li><code>NextGP.jl</code> relies on several other Julia packages, which could be easily avoided and may be removed in the later releases.</li><li>Currently one core dependency is the <a href="https://juliastats.org/StatsModels.jl/latest/"><code>StatsModels.jl</code></a> package, for model expression, and fixed effect design matrix generation.</li><li>Checking <code>StatsModels.jl</code>&#39;s manual for at least <a href="https://juliastats.org/StatsModels.jl/latest/formula/#The-@formula-language"><code>formula</code></a>  and  <a href="https://juliastats.org/StatsModels.jl/latest/contrasts/#Modeling-categorical-data"><code>categorical data</code></a> could be useful. </li></ul><hr/><h2 id="Basic-Model"><a class="docs-heading-anchor" href="#Basic-Model">Basic Model</a><a id="Basic-Model-1"></a><a class="docs-heading-anchor-permalink" href="#Basic-Model" title="Permalink"></a></h2><p><code>NextGP.jl</code> uses the following basic model:</p>$<p>\mathbf{y}=\mathbf{X}\mathbf{b}+\sum{\mathbf{Zu}}+\sum{\mathbf{M\beta}}+\mathbf{e} $</p><p><strong>b</strong> is a vector of fixed effects </p><p><strong>u</strong> is a vector of polygenic effects in the model with a covariance structure <span>$\mathbf{u}\sim N\left(0,\mathbf{A}\sigma_{a}^{2}\right)$</span></p><p><strong>A</strong> is a relationship matrix from the pedigree</p><p><strong>Z</strong> is a design matrix allocating animals to records</p><p><strong>M</strong> are matrices of genotypes</p><p class="math-container">\[\mathbf{\beta}\]</p><p>are vectors  of marker effects</p><hr/><h2 id="Public-Functions"><a class="docs-heading-anchor" href="#Public-Functions">Public Functions</a><a id="Public-Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Public-Functions" title="Permalink"></a></h2><p>Public functions of NextGP are documented here</p><article class="docstring"><header><a class="docstring-binding" id="NextGP.MCMC.runLMEM" href="#NextGP.MCMC.runLMEM"><code>NextGP.MCMC.runLMEM</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">function runLMEM(formula,userData,nChain,nBurn,nThin;myHints=Dict{Symbol,Any}(),blockThese=[],outFolder=&quot;outMCMC&quot;,VCV=[],userPedData=[],summaryStat=Dict{Any,Any}())</code></pre><ul><li><code>formula</code> is the model equatio as made available by the <a href="https://juliastats.org/StatsModels.jl/latest/"><code>StatsModels.jl</code></a> package</li><li><code>userData</code> is a <code>DataFrame</code> including <code>lhs</code> and <code>rhs</code> variables (other than defined by <code>PED</code> and <code>SNP</code>)</li><li><code>nChain</code>,<code>nBurn</code> and <code>nThin</code> are the chain length, burn-in period, and the thining interval used for the MCMC sampling</li><li>Users can define coding of their variables (e.g. full dummy coding) by providing <code>myHints</code>. Check <code>StatsModels.jl</code>&#39;s manual for <a href="https://juliastats.org/StatsModels.jl/latest/contrasts/#Modeling-categorical-data"><code>categorical data</code></a> could be useful.  </li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/datasciencetoolkit/NextGP.jl/blob/525f98b07499abea2ed4f55cb59f8b71baec7c3a/src/MCMC.jl#L23-L30">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="NextGP.makeA" href="#NextGP.makeA"><code>NextGP.makeA</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">makeA(s::Any, d::Any)</code></pre><p>Makes pedigree-based relationship matrix. adapted from http://morotalab.org/Mrode2005/relmat/createA.txt</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/datasciencetoolkit/NextGP.jl/blob/525f98b07499abea2ed4f55cb59f8b71baec7c3a/src/misc.jl#L67-L72">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="NextGP.makePed" href="#NextGP.makePed"><code>NextGP.makePed</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">makePed(inputFile::String,userData)</code></pre><p>Makes pedigree using PedigreeBase package</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/datasciencetoolkit/NextGP.jl/blob/525f98b07499abea2ed4f55cb59f8b71baec7c3a/src/misc.jl#L92-L97">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="NextGP.makeG" href="#NextGP.makeG"><code>NextGP.makeG</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">makeG(inputFile::String;method=1)</code></pre><p>Makes genomic relationship matrix based on vanRaden method 1 (defult) or method 2</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/datasciencetoolkit/NextGP.jl/blob/525f98b07499abea2ed4f55cb59f8b71baec7c3a/src/misc.jl#L118-L121">source</a></section><section><div><pre><code class="nohighlight hljs">    makeG(inputData::Array{Float64,2};method=1)</code></pre><p>Makes genomic relationship matrix based on vanRaden method 1 (defult) or method 2</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/datasciencetoolkit/NextGP.jl/blob/525f98b07499abea2ed4f55cb59f8b71baec7c3a/src/misc.jl#L141-L144">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="NextGP.SNP" href="#NextGP.SNP"><code>NextGP.SNP</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">    function SNP(name,path;map=&quot;&quot;)</code></pre><ul><li>Defines SNP information for further analysis.</li><li><code>path</code> is the path for the marker file, or the matrix for the marker genotypes.</li><li>Marker files are currently expected to be ordered as the phenotype data.</li><li>The method to be applied to the data, <code>GBLUP</code>, <code>BayesB</code>, <code>BayesR</code> etc, is defined in the prior setting.</li><li>Map file is optional. If not provided, a Bayesian Regression model with common variance for all SNPs will be applied. If provided, shoul match the order in the genotype file.</li><li>One most avoid overlapping marker sets by using different <code>name</code>s.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/datasciencetoolkit/NextGP.jl/blob/525f98b07499abea2ed4f55cb59f8b71baec7c3a/src/runTime.jl#L19-L27">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="NextGP.BayesPR" href="#NextGP.BayesPR"><code>NextGP.BayesPR</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">    function BayesPR(r,v)</code></pre><ul><li><code>r</code> is the region size. In other words, the number of SNPs that share a common variance.<ul><li><code>1</code>: each SNP has its own (co)variance</li><li><code>99</code>: SNPs on the same chromosome has the same (co)variance</li><li><code>9999</code>: All SNPs have the same (co)variance</li><li>One can define any other region size, for example, 30, 40 or 100</li></ul></li><li><code>v</code> is an estimate of the variance for the distribution of SNPs</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/datasciencetoolkit/NextGP.jl/blob/525f98b07499abea2ed4f55cb59f8b71baec7c3a/src/runTime.jl#L36-L44">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="NextGP.BayesB" href="#NextGP.BayesB"><code>NextGP.BayesB</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">    function BayesB(pi,v;estimatePi=false)</code></pre><ul><li><code>pi</code> is the proportion of SNPs to be included in the model at each McMC cycel. If <code>estimatePi=true</code>, it is only used as a starting value.</li><li><code>v</code> is the variance for the prior distribution of SNPs.</li><li><code>estimatePi</code> is <code>true</code>if <code>pi</code> is estimated. By default it is ´false´</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/datasciencetoolkit/NextGP.jl/blob/525f98b07499abea2ed4f55cb59f8b71baec7c3a/src/runTime.jl#L55-L60">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="NextGP.BayesC" href="#NextGP.BayesC"><code>NextGP.BayesC</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">    function BayesC(pi,v;estimatePi=false)</code></pre><ul><li><code>pi</code> is the proportion of SNPs to be included in the model at each McMC cycel. If <code>estimatePi=true</code>, it is only used as a starting value.</li><li><code>v</code> is the variance for the prior distribution of SNPs.</li><li><code>estimatePi</code> is <code>true</code>if <code>pi</code> is estimated. By default it is ´false´</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/datasciencetoolkit/NextGP.jl/blob/525f98b07499abea2ed4f55cb59f8b71baec7c3a/src/runTime.jl#L70-L75">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="NextGP.BayesR" href="#NextGP.BayesR"><code>NextGP.BayesR</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">    function BayesR(pi,class,v;estimatePi=false)</code></pre><ul><li><code>pi</code> is the vector of proportion of SNPs for each variance class. If <code>estimatePi=true</code>, it is only used as a starting value.</li><li><code>class</code> is the vector of scales of common SPN variance for each variance class. The scales should be in the increasing order. For example, [0.0,0.0001,0.001,0.01].</li><li><code>v</code> is the variance for the prior distribution of SNPs.</li><li><code>estimatePi</code> is <code>true</code>if <code>pi</code> is estimated. By default it is ´false´</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/datasciencetoolkit/NextGP.jl/blob/525f98b07499abea2ed4f55cb59f8b71baec7c3a/src/runTime.jl#L86-L92">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="NextGP.BayesLV" href="#NextGP.BayesLV"><code>NextGP.BayesLV</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">    function BayesLV(v,f,covariates,zeta)</code></pre><ul><li><code>v</code> is the variance for the prior distribution of SNPs.</li><li><code>f</code> is the model formula for the variance</li><li><code>covariates</code>is the <code>DataFrame</code> that includes explanatory varibles for the variance of each SNP.</li><li><code>zeta</code> is the error variance in the model for the SNP variances.</li><li>If <code>estimateVarZeta</code> is <code>true</code>, it assumes that the error variance for the model for the SNP variances is 0.01 percent of the variance of the log-variances. If <code>estimateVarZeta</code> is <code>false</code>, it uses <code>zeta</code> as the error variance for the log-linear variances (fixed value). If <code>estimateVarZeta</code> is a <code>Float64</code>, it assumes that the error variance for the model for the SNP variances is <code>estimateVarZeta</code> percent of the variance of the log-variances.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/datasciencetoolkit/NextGP.jl/blob/525f98b07499abea2ed4f55cb59f8b71baec7c3a/src/runTime.jl#L125-L132">source</a></section></article><h2 id="Internals"><a class="docs-heading-anchor" href="#Internals">Internals</a><a id="Internals-1"></a><a class="docs-heading-anchor-permalink" href="#Internals" title="Permalink"></a></h2><p>Functions that are used internally are documented here</p><article class="docstring"><header><a class="docstring-binding" id="NextGP.prepMatVec.prep" href="#NextGP.prepMatVec.prep"><code>NextGP.prepMatVec.prep</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">function prep(f::StatsModels.TermOrTerms, inputData::DataFrame;userHints=Dict{Symbol,Any}(),path2ped=[],priorVCV=[])</code></pre><ul><li><code>NextGP</code> relies on <code>StatsModels.jl</code> package for model expression (<code>f</code>), and fixed effect design matrix generation.</li><li>Details for the model expression (<code>f</code>), and fixed effects coding specifications (e.g., effect or dummy coding) can be found at <a href="https://juliastats.org/StatsModels.jl/latest/"><code>StatsModels.jl</code></a>.</li><li>Design matrices for random effects are generated either own internal functions or using <code>StatsModels.jl</code>s <code>modelcols</code>, depending on how user defined the model term in the model.</li><li>Reads in marker data, and mean-centers the columns.</li><li>Finally returns lhs vector and rhs matrices.</li><li>By default:<ul><li>all <code>Int</code> rhs variables are made <code>Categorical</code>,</li><li>all <code>String</code> rhs variables (also those made <code>Categorical</code>) are dummy coded, except those defined by the user in <code>userHints</code>, </li><li>all <code>Float</code> rhs variables are centered.</li></ul></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/datasciencetoolkit/NextGP.jl/blob/525f98b07499abea2ed4f55cb59f8b71baec7c3a/src/prepMatVec.jl#L13-L25">source</a></section></article><h2 id="Convenience-functions"><a class="docs-heading-anchor" href="#Convenience-functions">Convenience functions</a><a id="Convenience-functions-1"></a><a class="docs-heading-anchor-permalink" href="#Convenience-functions" title="Permalink"></a></h2><h3 id="Convergency-statistics-of-parameter-param,-using-MCMCChains.jl"><a class="docs-heading-anchor" href="#Convergency-statistics-of-parameter-param,-using-MCMCChains.jl">Convergency statistics of parameter <code>param</code>, using <code>MCMCChains.jl</code></a><a id="Convergency-statistics-of-parameter-param,-using-MCMCChains.jl-1"></a><a class="docs-heading-anchor-permalink" href="#Convergency-statistics-of-parameter-param,-using-MCMCChains.jl" title="Permalink"></a></h3><pre><code class="language-julia hljs">using MCMCChains,StatsPlots,StatsBase,CSV

function convergencyMCMC(param;summary=false,plots=false,outFolder=pwd()*&quot;/outMCMC&quot;)
        param = CSV.read(&quot;$outFolder/$(param)Out&quot;,DataFrame,header=true)
        namesParam = names(param)
        param = Matrix(param)
                if summary==true
                        chn = Chains(param,namesParam)
                        display(chn)
                        if plots==true
                                display(plot(chn))
                        end
                        param = mean(Matrix(param),dims=1)
                else
                        param = mean(Matrix(param),dims=1)
                end
        return param
end
</code></pre><ul><li>If <code>summary=true</code>, will print convergency statistics for McMC</li><li>If <code>plots=true</code>, will print trace plot(s) of McMC</li><li><code>outFolder</code> is the folder for the McMC output. By default it searches for the folder &quot;outMCMC&quot; in the current directory.</li></ul><hr/></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="home/">« Home</a><a class="docs-footer-nextpage" href="Example/Example/">Example »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Tuesday 1 October 2024 11:19">Tuesday 1 October 2024</span>. Using Julia version 1.7.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
